feat(persist): implement persist in v1000

osovv opened the pull request.

osovv committed:
feat(core): add persist adapter for state persistence

osovv committed:
feat(core): add web storage persist adapters

osovv committed:
docs(persist): completely rewrite documentation with web storage adapters

osovv committed:
refactor(persist): improve web storage adapter

osovv committed:
feat(examples): add react persist web storage example

artalar commented on examples/react-persist-web/src/app.tsx:
```diff
@@ -0,0 +1,170 @@
+} from '@reatom/core'
+
+// Create BroadcastChannel for demo
+const demoChannel =
```
~~Where is it possible? Why we need it for the demo? Lets use the defaults~~

Ok, lets use it as the configuration example, but add a little more comments

artalar commented on examples/react-persist-web/src/app.tsx:
```diff
@@ -0,0 +1,170 @@
+// Storage demo component
+const StorageDemo = reatomComponent<{
+  title: string
+  atom: typeof localStorageAtom
```
```suggestion
  model: typeof localStorageAtom
```

artalar commented on examples/react-persist-web/src/app.tsx:
```diff
@@ -0,0 +1,170 @@
+    : null
+
+// Create atoms with different storage adapters
+const localStorageAtom = atom('', 'localStorageValue').extend(
```
And so oon
```suggestion
const dataWithLocalStorage = atom('', 'localStorageValue').extend(
```

artalar commented on examples/react-persist-web/src/app.tsx:
```diff
@@ -0,0 +1,170 @@
+  return (
+    <div style={{ padding: '20px', maxWidth: '800px', margin: '0 auto' }}>
+      <h1>🔄 Reatom Persist Demo</h1>
+      <p>Try different storage adapters and see how data persists!</p>
```
```suggestion
      <p>Try different storage adapters and see how data persists!</p>
      <p>Open a few tabs and see the realtime sync.</p>
      <p>Open the devtools Application tab and inspect the data.</p>
```

artalar commented on packages/core/src/persist/index.ts:
```diff
@@ -0,0 +1,336 @@
+
+export interface PersistRecord<T = unknown> {
+  data: T
+  /** @deprecated Not needed anymore */
```
;)

artalar commented on packages/core/src/persist/index.ts:
```diff
@@ -0,0 +1,336 @@
+              const isWrite = params.length > 0
+
+              // Initialize from storage on first read
+              if (!isInitialized && !isWrite) {
```
Use `isInit()`

artalar commented on packages/core/src/persist/index.ts:
```diff
@@ -0,0 +1,336 @@
+                  // Handle both sync and async results
+                  if (result instanceof Promise) {
+                    // Async storage
+                    wrap(async () => {
```
No need `wrap` there as you didn't perform any computations after than

artalar commented on packages/core/src/persist/index.ts:
```diff
@@ -0,0 +1,336 @@
+                      } catch (error) {
+                        console.warn('Failed to load persisted state:', error)
+                      }
+                    })().catch(console.warn)
```
Leave it "unhandled"

artalar commented on packages/core/src/persist/index.ts:
```diff
@@ -0,0 +1,336 @@
+
+      // Add storage subscription if enabled (always active)
+      if (subscribe && storage.subscribe) {
+        storage.subscribe(key, () => {
```
withConnectHook?

artalar commented on packages/core/src/persist/index.ts:
```diff
@@ -0,0 +1,336 @@
+            // Handle both sync and async results
+            if (result instanceof Promise) {
+              result
+                .then((newRecord) => {
```
```suggestion
                .then(wrap((newRecord) => {
```

artalar commented on packages/core/src/persist/index.ts:
```diff
@@ -0,0 +1,336 @@
+      // Notify listeners
+      const listeners = listenersAtom().get(key)
+      if (listeners) {
+        listeners.forEach((cb) => cb(rec))
```
Add `_enqueue(, 'effect')`

osovv committed:
fix(persist): resolve storage subscription and PR review comments

Addresses all 11 comments from PR #1152:

Core persist implementation fixes:
- Replace manual isInitialized flag with proper isInit() helper
- Remove deprecated comment marker and unnecessary wrap() calls
- Remove .catch(console.warn) to leave errors unhandled as requested
- Fix withConnectHook usage to pass target parameter correctly
- Use _enqueue with 'effect' queue for subscription callbacks
- Add value-based comparison instead of ID-based for subscriptions

React example improvements:
- Enhance BroadcastChannel comments explaining cross-tab sync
- Rename atoms from localStorageAtom to dataWithLocalStorage for clarity
- Fix prop type from atom to model with proper Atom<string> typing
- Add detailed usage instructions and demo guidance

Test fixes:
- Fix async withConnectHook activation with wrap(sleep()) pattern
- Add proper subscription tracking and notify() calls for effect queue
- Ensure subscription system works for multiple atoms sharing storage keys

osovv commented on packages/core/src/persist/index.ts:
```diff
@@ -146,14 +150,14 @@ export const reatomPersist = (
                   if (result instanceof Promise) {
                     // Async storage
-                    wrap(async () => {
+                    ;(async () => {
```
this `;` comes from prettier
